package/MDAnalysis/lib/pkdtree.py

```diff
@@ -0,0 +1,181 @@
+
+from __future__ import absolute_import
+
+import numpy as np
+from Bio.KDTree import _CKDTree
+
+__all__ = ['PeriodicKDTree', ]
+
+
+class PeriodicKDTree(object):
+    """
+    Wrapper around Bio.KDTree._CKDTree to enable search with periodic boundary
+     conditions.
+
+    A tree is first constructed with the coordinates wrapped onto the central
+    cell. A query for neighbors around a center point is performed first by
+    wrapping the center point coordinates to the central cell, then generating
+    images of this wrapped center point and finally searching for neighbors
+    close to the images.
+
+    Only the necessary number of center point images is generated for each
+    case. For instance, if the wrapped center point lies well within the cell
+    and far away from the cell boundaries, there may be no need to generate
+    any center point image.
+    """
+
+    def __init__(self, box, bucket_size=10):
+        """
+        Parameters
+        ----------
+        box : array-like or ``None``, optional, default ``None``
+          Simulation cell dimensions in the form of
+          :attr:`MDAnalysis.trajectory.base.Timestep.dimensions` when
+          periodic boundary conditions should be taken into account for
+          the calculation of contacts.
+        bucket_size : int
+          Number of entries in leafs of the KDTree. If you suffer poor
+          performance you can play around with this number. Increasing the
+          `bucket_size` will speed up the construction of the KDTree but
+          slow down the search.
+        """
+        self.dim = 3  # Strict implementation for 3D
+        self.kdt = _CKDTree.KDTree(self.dim, bucket_size)
+        self.built = 0
+        if len(box) != self.dim:
+            raise Exception('Expected array of length {}'.format(self.dim))
+        self.box = np.copy(np.asarray(box))
+        # zero is the flag for no periodic boundary conditions
+        self.box = np.where(self.box == np.inf, 0.0, self.box)
+        self.box = np.where(self.box > 0.0, self.box, 0.0)
+        if not self.box.any():
+            raise Exception('No periodic axis found')
+        self._indices = list()
+
+    def set_coords(self, coords):
+        """
+        Add coordinates of the points. Wrapping of coordinates to the central
+        cell is enforced along the periodic axes.
+
+        Parameters
+        ----------
+        coords: NumPy.array
+          Positions of points, shape=(N, 3) for N atoms.
+        """
+        if coords.min() <= -1e6 or coords.max() >= 1e6:
+            # Same exception class as in Bio.KDTree.KDTree.set_coords
+            raise Exception('Points should lie between -1e6 and 1e6')
+        if len(coords.shape) != 2 or coords.shape[1] != self.dim:
+            raise Exception('Expected a (N, {}) NumPy array'.format(self.dim))
+        wrapped_data = (coords - np.where(self.box > 0.0,
+                                          np.floor(coords / self.box) * self.box, 0.0))
```

Comment:
This will only work for cubic boxes.

Reply:
@kain88-de As far as I understand, `Bio.KDTree._cKDTree` assumes that coordinates are expressed with respect to a set of orthonormal axes. This has to be the case because the search for neighbors within a sphere of a certain radius implies an isotropic space. So, if one loads the PDB file of an orthorhombic system, the wrapping will still work because PDB coordinates are expressed all in the same absolute units (Angstroms), that is, they are not fractional coordinates.
The wrapping will not work for the general triclinic system, although imaging for the triclinic system is easy to implement. The critical point is that coordinates must be expressed with respect to an orthonormal set of axes. I guess my question is how can one make sure this is the case?

Reply:
You can convert the box to a triclinic representation. The box vectors should be scaled to 1 to they are unit vectors. From that matrix the determinant should be 1 to have a orthonormal coordinate system. 

But mostly we need to document this.

Reply:
[We've got some code that does this transformation (incl. triclinic)](https://github.com/MDAnalysis/mdanalysis/blob/develop/package/MDAnalysis/lib/distances.py#L774).  I'd rather reuse that then have different implementations all over the code base.

We only really officially support orthogonal and triclinic boxes (ie can be represented as `x, y, z, alpha, beta, gamma`)

Reply:
Although admittedly that function doesn't support non periodic dimensions.  Maybe should raise an issue making `x=0` indicate that the system is non periodic in that dimension

Reply:
I will use [apply_PBC](https://github.com/MDAnalysis/mdanalysis/blob/develop/package/MDAnalysis/lib/distances.py#L774) and leave out for the moment semi-periodic systems. These systems could be later implemented by refactoring [_ortho_pbc](https://github.com/MDAnalysis/mdanalysis/blob/develop/package/MDAnalysis/lib/include/calc_distances.h#L85) and [_triclinic_pbc](https://github.com/MDAnalysis/mdanalysis/blob/develop/package/MDAnalysis/lib/include/calc_distances.h#L102)